Enable -Wunsafe-buffer-usage-in-libc-call

geoffreygaren opened the pull request:
#### 637c4dc64e64ad103d707b5192a3b48050efbac6
<pre>
Enable -Wunsafe-buffer-usage-in-libc-call
<a href="https://bugs.webkit.org/show_bug.cgi?id=287209">https://bugs.webkit.org/show_bug.cgi?id=287209</a>
<a href="https://rdar.apple.com/144351483">rdar://144351483</a>

Reviewed by David Kilzer.

Only very recent compilers enforce this warning.

I built locally with one of them and it seems to pass.

Also fixed up some cases I missed before.

* Configurations/CommonBase.xcconfig: I&apos;m leaving behind some of the .xcconfig
support for disabling this warning for now, just in case.

* Source/WTF/wtf/Compiler.h: Skip the new warning when skipping warnings.

* Source/WTF/wtf/StdLibExtras.h: Put our skip markers inside our macro because
markers around the macro only seem to take effect when defining the macro,
and not when expanding it.

* Configurations/CommonBase.xcconfig: Enable.

Canonical link: <a href="https://commits.webkit.org/290060@main">https://commits.webkit.org/290060@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a2f5451a9a92c82330c7c40eea2aa8c3e21370ce

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/88852 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8376 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43320 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/93822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/39611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/90903 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/8763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16560 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/93822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/39611 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/91854 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/8763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/80346 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/93822 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/8763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/34760 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe-cairo~~](https://ews-build.webkit.org/#/builders/65/builds/38719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/81651 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/8763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/35660 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95661 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/87628 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/16032 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/11719 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/77351 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/16288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76212 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/76636 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19448 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/9098 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/13909 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/16046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/21357 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/110121 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/15787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/26432 "Passed tests") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/19238 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/17568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WTF/wtf/Compiler.h:
```diff
@@ -619,6 +619,8 @@
 #define WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN \
     _Pragma("clang diagnostic push") \
     _Pragma("clang diagnostic ignored \"-Wunsafe-buffer-usage\"")
+    _Pragma("clang diagnostic ignored \"-Wunknown-warning-option\"")
```
Maybe put this one before `-Wunsafe-buffer-usage`?

ddkilzer reviewed (approved):
r=me, but consider putting `-Wunknown-warning-option` first in the pragma list.

ddkilzer commented on Source/WebCore/loader/FTPDirectoryParser.cpp:
```diff
@@ -146,7 +146,9 @@ FTPEntryType parseOneFTPLine(std::span<LChar> line, ListState& state, ListResult
                                 pos++;
                             if (pos < linelen && line[pos] == ',') {
                                 unsigned long long seconds = 0;
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
                                 sscanf(byteCast<char>(p.subspan(1)).data(), "%llu", &seconds);
+WTF_ALLOW_UNSAFE_BUFFER_USAGE_END
```
This source is supposed to be dead code.  Seems like it's still compiled, but should be removable.

geoffreygaren replied:
@beidson @cdumez Can we delete this code?

ddkilzer reviewed (approved):
r=me.  FTP code can be removed in a separate PR.

geoffreygaren committed:
Enable -Wunsafe-buffer-usage-in-libc-call
https://bugs.webkit.org/show_bug.cgi?id=287209
rdar://144351483

Reviewed by David Kilzer.

Only very recent compilers enforce this warning.

I built locally with one of them and it seems to pass.

Also fixed up some cases I missed before.

* Configurations/CommonBase.xcconfig: I'm leaving behind some of the .xcconfig
support for disabling this warning for now, just in case.

* Source/WTF/wtf/Compiler.h: Skip the new warning when skipping warnings.

* Source/WTF/wtf/StdLibExtras.h: Put our skip markers inside our macro because
markers around the macro only seem to take effect when defining the macro,
and not when expanding it.

* Configurations/CommonBase.xcconfig: Enable.

Canonical link: https://commits.webkit.org/290060@main

cdumez commented on Source/WebCore/bridge/objc/objc_class.mm:
```diff
@@ -133,7 +133,7 @@ static inline void convertJSMethodNameToObjc(const CString& jsName, JSNameConver
                 mappedName = [thisClass webScriptNameForSelector:objcMethodSelector];
 
-            if ((mappedName && [mappedName isEqual:methodName.get()]) || !strcmp(objcMethodSelectorName, buffer.data())) {
+            if ((mappedName && [mappedName isEqual:methodName.get()]) || equalSpans(objcMethodSelectorName, buffer.span())) {
```
This is not quite right and broke WK1 tests on mac-wk1. Fixing in https://github.com/WebKit/WebKit/pull/40334.
